File: ACM/Acwing/1460.py
'''
Author: NEFU AB-IN
Date: 2022-02-14 18:15:29
FilePath: \ACM\Acwing\1460.py
LastEditTime: 2022-02-14 20:16:51
'''

from collections import Counter

N = 155
P, MOD = 131, 1 << 64

p = [0] * N
h = [0] * N

p[0] = 1

d = Counter()


def get(l, r):
    return (h[r] - h[l - 1] * p[r - l + 1]) % MOD


def check(x):
    for i in range(1, n + 1):
        if i + x - 1 > n:
            break
        ss = get(i, i + x - 1)
        if d[ss]:
            return True
        d[ss] += 1
    return False


if __name__ == "__main__":
    n = int(input())
    s = input()

    s = " " + s
    for i in range(1, n + 1):
        p[i] = p[i - 1] * P % MOD
        h[i] = (h[i - 1] * P + ord(s[i])) % MOD

    l = 1
    r = n
    while l < r:
        mid = l + r + 1 >> 1
        if check(mid):
            l = mid
        else:
            r = mid - 1
    print(r + 1)

File: ACM/Acwing/1460/1460.cpp
/*
 * @Author: NEFU AB-IN
 * @Date: 2023-02-15 15:03:27
 * @FilePath: \Acwing\1460\1460.cpp
 * @LastEditTime: 2023-02-15 15:32:33
 */
#include <bits/stdc++.h>
using namespace std;
#define int long long
#undef int

#define SZ(X) ((int)(X).size())
#define ALL(X) (X).begin(), (X).end()
#define IOS                                                                                                            \
    ios::sync_with_stdio(false);                                                                                       \
    cin.tie(nullptr);                                                                                                  \
    cout.tie(nullptr)
#define DEBUG(X) cout << #X << ": " << X << '\n'
typedef pair<int, int> PII;

const int N = 1e5 + 10, INF = 0x3f3f3f3f;

typedef unsigned long long ULL;
ULL p[N], h[N];
const int P = 13331;
ULL get(int l, int r)
{
    return h[r] - h[l - 1] * p[r - l + 1];
}

int n;
string s;

bool check(int x)
{
    unordered_map<int, int> st;
    for (int i = 1; i + x - 1 <= n; ++i)
    {
        int ss = get(i, i + x - 1);
        if (st.count(ss))
            return true;
        st[ss]++;
    }
    return false;
}

signed main()
{
    IOS;
    cin >> n >> s;
    s = " " + s;
    p[0] = 1;
    for (int i = 1; i <= n; ++i)
    {
        p[i] = p[i - 1] * P;
        h[i] = h[i - 1] * P + s[i];
    }

    int l = 1, r = n;
    while (l < r)
    {
        int mid = (l + r) >> 1;
        if (check(mid))
            r = mid;
        else
            l = mid + 1;
    }
    cout << l;
    return 0;
}